Make `around` types based off the mapped properties of the object being patched, not a generic index signature.

just-be-dev opened the pull request:
**Note: This is technically a breaking change for the types**

## The problem

I'm trying to monkey-patch two functions on Obsidian's publish plugin's instance. 

`apiRequest` and `apiUploadFile`. These functions have different signatures which I've defined. In defining those signatures, I've been getting TypeScript errors related to the index signature. 

```
(property) apiUploadFile: (baseApiUploadFile: APIUploadFn) => APIUploadFn
Type '(baseApiUploadFile: APIUploadFn) => APIUploadFn' is not assignable to type 'methodWrapperFactory<(options: UploadArgs) => Promise<any>>'.
  Types of parameters 'baseApiUploadFile' and 'next' are incompatible.
    Types of parameters 'options' and 'file' are incompatible.
      Type 'TAbstractFile' is missing the following properties from type 'UploadArgs': method, url, headers, datats(2322)
index.d.ts(4, 5): The expected type comes from this index signature.
```

There's a lot going on there and without looking at my code it's likely not clear. One of the functions is defined to take `UploadArgs` and the other takes `TAbstractFile`. I believe given how the current type signature is working, it's inferring that `methodWrapperFactory<T>` is a singular definition and thus requires the same signatures for each of the functions. 

## My changes

What I've done instead is convert this to a mapped type where it'll infer the type of the Object passed and reference it's direct children for the `methodWrapperFactory` values. 

## Other impacts

As mentioned previously, this is breaking because it changes `around`'s type signature pretty significantly. It's now more sound but also more strict. So long as the thing you're monkey patching has good types, it's great! Forces adherence to those types all the way down (unless you specifically cast `around`). I do get that a lot of things that are being monkey patched aren't typed because they're usually private APIs. That adds the burden of adding extra types, but is worth it in my mind.

just-be-dev committed:
Add mapped type to avoid inferring T as a union

just-be-dev committed:
Further tweaking to allow functions not in O

just-be-dev committed:
Directly assume method wrapper is exact mapped property of O

pjeby commented:
I'm not sure I like needing to add extra types for patching private APIs, but then again I mostly use around() from vanilla JS anyway.

In principle, I'm not against adding this, but don't know enough to be sure I won't turn around in a month or two and realize it was a mistake and needing to break compatibility again switching it back.  I think what I'll need to do is try the patch myself and see what it does in my projects first, and I'm not sure when that will be.  I assume, though, that you can depend on your own fork in the meantime?

just-be-dev commented:
Yeah, there's no rush from my side. I'm using [patch-package](https://www.npmjs.com/package/patch-package) so it's not blocking me.

just-be-dev commented:
You can always break out of the type checking if you need to. `around<any>(...)` will effectively disable the type checking. Or really you could just include a union type:

```ts
around<PublicAPI & PrivateAPI>(...)
```

In my mind it's a bit preferable just because the typesafety is what we're going after. If you use it think it hinders you though, I get that too. No pressure either way.

pjeby commented:
I'm more of a "types are documentation" kind of person than "types are safety", really.  The main reason this has types at all was just to make things slightly easier for folks using TS or VSCode and serve as an alternative to more verbose documentation.

Not real clear on why you need patch-package, since you could just reference your fork directly.  Also, now that I'm looking at it more closely, I'm not clear on why you patched the .d.ts instead of the source index.ts?

just-be-dev commented:
🤦 

Because sometimes as detailed as I try to be I'm incredibly unobservant. I missed that.

I'll update the PR this week. 

(Generally I prefer patch-package for things like slightly changing types because I don't want to assume the maintenance and burden of drift when a package updates. If the patched file hasn't changed then no updates will be needed. If it has, it's easy enough to correct that. Easier than using a fork imo)

pjeby commented:
I'm also kind of wondering if there's any way to constrain that the keys being affected have to be functions?  Technically the first argument can accept stuff besides functions, but the second argument has to have methods that accept and return functions, not arbitrary properties.  (Also, I expect the typing of `around1` will need to change in a compatible fashion as well.)

just-be-dev committed:
Update index types to match type definition file

pjeby commented:
Any progress on this?  It currently looks like the wrapper type is being treated as `any`, which isn't what we want, I don't think.

just-be-dev committed:
Allow factory to provide partial results, accept only functions

Revert filter change

Revert changes to original state

Update output

Re-add function extension

just-be-dev commented:
Okay, I pushed an update. I've done three things:
1. I made `factories` partial meaning you can mock out a sub-set of a defined object which is probably more of what folks would want.
2. I made the keys be constrained to functions. There's a caveat here in that since the overall type accepts a partial now technically `undefined` still passes type checking. 
3. Update the runtime to only accept functions

For example if you did this...

```ts
// assume myObject has a `test` proprety of `boolean`
around(myObject, {
  test: true
})
```
☝️ that would correctly fail

```ts
// assume myObject has a `test` proprety of `boolean`
around(myObject, {
  test: undefined
})
```
☝️ that would actually pass type checking. 

I added a runtime check to also ensure that only functions are processed. This solution is imperfect. From a types perspective it'd be nice if it correctly failed on any non-function and silently stripping out non-functions at runtime doesn't actually give the user a lot of details. Perhaps it should error instead?

Open to thoughts.

pjeby commented:
I don't see the point to a runtime check -- *not* checking would throw an error, so it's actually better *not* to do the filter.

I'm also not clear on the purpose of the change to MethodWrapperFactory.  It doesn't seem to be adding anything.

At the moment, I'm leaning to just changing the typings on around and around1 to match what you have here, but nothing else.  Would that work for you, or is there something more you need?

just-be-dev commented:
Okay, I cleaned things up and I think it's closer to where you wanted it to be now. Thanks for bearing with me!

You'd mentioned `wrapper` being casted as any.

`  (obj[method] as any) = wrapper;`

Technically at this point `obj[method]` is typed as `O[keyof O]` which is technically correct. It's a mapping to a method on the object. TypeScript isn't broadening out the definition to encapsulate everything in the `Function` space as we specified in the `methodWrapperFactory`. I'm not exactly sure _why_ this is happening. Generally though, given that `wrapper` is an internal aspect of the API, I think it's safe to cast here for the purpose of reassignment.

pjeby commented:
Yep.  I've gone ahead and release 2.2.0 with `obj[method] = wrapper as O[keyof O]`. as it's more clear that the wrapper is generic in this context.  I'm assuming this should be sufficient for you?  I just used the library in TypeScript for the first time and the new typing worked where the old one did not.

just-be-dev committed:
Cast wrapper type

just-be-dev commented:
I added that. If there's anything you want to edit in this PR, feel free.

pjeby committed:
Merge branch 'master' into master

pjeby commented:
I already released 2.2.0 with my version of this?  (See previous comment.)  Looks like I forgot to push my commit, though; see a23dd0227ae531b7d7ea02885a4652baf036ff1e.  (In fact I forgot to push the tag for my *previous* release, from back in March!)  Sorry for the confusion.
